conbencher/python/conbencher/adapters/_adapter.py

```diff
@@ -0,0 +1,86 @@
+import abc
+import subprocess
+
+from ..client import ConbenchClient
+from ..log import fatal_and_log
+from ..result import BenchmarkResult
+
+
+class _BenchmarkAdapter(abc.ABC):
+    """
+    An abstract class to run benchmarks, transform results into conbench form,
+    and send them to a conbench server
+
+    Attributes
+    ----------
+    command : list[str]
+        A list of args to be run on the command line, as would be passed
+        to `subprocess.run()`.
+    results : list[BenchmarkResult]
+        Once `run()` has been called, results from that run
+    """
+
+    command: list[str]
+    results: list[BenchmarkResult] = None
+
+    def __init__(self, command: list[str]) -> None:
+        self.command = command
+
+    def __call__(self, **kwargs) -> list:
+        """
+        Run benchmarks and post results to conbench server
+
+        Parameters
+        ----------
+        kwargs
+            Passed through to `run()`
+        """
+        self.run(**kwargs)
+        self.post_results()
+
+    def run(self, params: list[str] = None) -> list[BenchmarkResult]:
+        """
+        Run benchmarks
+
+        Parameters
+        ----------
+        params : list[str]
+            Additional parameters to be appended to the command before running
+        """
+        command = self.command
+        if params:
+            command += params
+
+        res = subprocess.run(args=command)
+        res.check_returncode()
+
+        self.results = self.transform_results()
+
+        return self.results
+
+    @abc.abstractmethod
+    def transform_results(self) -> list[BenchmarkResult]:
+        """
+        Method to transform results from the command line call into a list of
+        instances of `BenchmarkResult`.
+        """
+        raise NotImplementedError
```

Comment:
nit: Since this is an `abstractmethod` you don't need this line; the interpreter will throw an error when constructing a child class if it's not implemented. Usually I just have the docstring and that's enough to satisfy the syntax.

(this doesn't matter unless you're striving for 100% coverage someday 🚀)